vcr/statuslist/store.go

```diff
@@ -0,0 +1,287 @@
+			// see the "optimistic locking" section in tests
+			if errors.Is(tx.Error, gorm.ErrRecordNotFound) {
+				// gorm.ErrRecordNotFound means that the LastIssuedIndex has already been updated and a new index should be generated
+				log.Logger().Debug("StatusListIndex changed while updating. Generating new index.")
+				continue
+			}
+			return nil, tx.Error
+		}
+		// successfully incremented LastIssuedIndex
+		break
+	}
+
+	return &credential.StatusList2021Entry{
+		ID:                   fmt.Sprintf("%s#%d", credentialRecord.StatusListCredential, credentialRecord.LastIssuedIndex),
+		Type:                 credential.StatusList2021EntryType,
+		StatusPurpose:        StatusPurposeRevocation,
+		StatusListIndex:      strconv.Itoa(credentialRecord.LastIssuedIndex),
+		StatusListCredential: credentialRecord.StatusListCredential,
+	}, nil
+}
+
+func (s *sqlStore) Revoke(ctx context.Context, credentialID ssi.URI, entry credential.StatusList2021Entry) error {
+	// parse StatusListIndex
+	statusListIndex, err := strconv.Atoi(entry.StatusListIndex)
+	if err != nil {
+		return err
+	}
+
+	// validate StatusPurpose
+	if entry.StatusPurpose != StatusPurposeRevocation {
+		// cannot be issued by this node, so DB-free check if managed by this node
+		return errUnsupportedPurpose
+	}
+
+	// check if StatusList2021Credential is managed by this node
+	var statuslist statusListCredentialRecord
+	// SELECT * FROM status_list_credentials WHERE status_list_credential = 'entry.StatusListCredential' LIMIT 1;
+	err = s.DB(ctx).Where("status_list_credential = ?", entry.StatusListCredential).First(&statuslist).Error
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return errNotFound // statusListCredential not managed by this node
+		}
+		return err
+	}
+
+	// validate StatusListIndex
+	if statusListIndex < 0 || statusListIndex > statuslist.LastIssuedIndex {
+		return statuslist2021.ErrIndexNotInBitstring
+	}
+
+	// revoke
+	revocation := revocationRecord{
+		StatusListCredential: statuslist.StatusListCredential,
+		StatusListIndex:      statusListIndex,
+		CredentialID:         credentialID.String(),
+	}
+	// INSERT INTO status_list_statuses (status_list_credential, status_list_index, credentialID)
+	// VALUES ('statuslist.StatusListCredential', 'statusListIndex', 'credentialID.String()');
+	err = s.DB(ctx).Create(revocation).Error
+	if err != nil {
+		if errors.Is(err, gorm.ErrDuplicatedKey) {
+			return errRevoked // already revoked
+		}
+		return err
+	}
+
+	// successful revocation
+	return nil
+}
+
+func toStatusListCredential(issuer did.DID, page int) (string, error) {
+	switch issuer.Method {
+	case "web":
+		issuerAsURL, err := didweb.DIDToURL(issuer) // https://example.com/iam/id
+		if err != nil {
+			return "", err
+		}
+		return issuerAsURL.JoinPath("statuslist", strconv.Itoa(page)).String(), nil // https://example.com/iam/id/statuslist/page
+	}
+	return "", fmt.Errorf("status list: unsupported DID method: %s", issuer.Method)
```

Comment:
`DIDToURL` also returns this error, so you can skip the switch statement altogether

Reply:
`DIDToURL` joins two errors resulting in `URL does not represent a Web DID\nunsupported DID method: nuts`. This doesn't make much sense to me; even less when I wrap it in `status list: %w`

We could probably use generic `unsupported DID method` and `not supported for did method` errors to prevent this